qutip/solver/integrator.py

```diff
@@ -0,0 +1,185 @@
+
+        Return
+        ------
+        (t, state) : (float, qutip.Data)
+            The state of the solver at ``t``. The returned this can differ from
+            the input time only when ``step=True``.
+        """
+        raise NotImplementedError
+
+    def get_state(self, copy=True):
+        """
+        Obtain the state of the solver as a pair (t, state).
+
+        Return
+        ------
+        (t, state) : (float, qutip.Data)
+            The state of the solver at ``t``.
+        """
+        raise NotImplementedError
+
+    def run(self, tlist):
+        """
+        Integrate the system yielding the state for each times in tlist.
+
+        Parameters
+        ----------
+        tlist : *list* / *array*
+            List of times to yield the state.
+
+        Yield
+        -----
+        (t, state) : (float, qutip.Data)
+            The state of the solver at each ``t`` of tlist.
+        """
+        for t in tlist[1:]:
+            yield self.integrate(t, False)
+
+    def reset(self):
+        """Reset internal state of the ODE solver."""
+        if self._is_set:
+            self.set_state(*self.get_state())
+
+    def update_args(self, args):
+        """
+        Change the argument of the system.
+        Reset the ODE solver to ensure numerical validity.
+
+        Parameters
+        ----------
+        args : dict
+            New arguments
+        """
+        self.system.arguments(args)
+        self.reset()
+
+
+# TODO: These integrator set will be part of the solver classes
+# So sesolve will know which integrator it support etc.
+# `mesolve` and `sesolve` will use the same integrators, but some integrations
+# methods can not work for `mcsolve`.
+sesolve_integrators = {}
+mesolve_integrators = {}
+mcsolve_integrators = {}
+
+
+def add_integrator(integrator, keys, integrator_set, options_class=None):
+    """
+    Register an integrator.
+
+    Parameters
+    ----------
+    integrator : Integrator
+        The ODE solver to register.
+
+    keys : list of str
+        Values of the method options that refer to this integrator.
+
+    integrator_set : dict
+        Group of integrators to which register the integrator.
+
+    options_class : SolverOptions
+        If given, will add the ODE options supported by this integrator to
+        those supported by the options_class. ie. If the integrator use a
+        `stiff` options that Qutip's `SolverOdeOptions` does not have, it will
+        add it support.
+    """
+    # TODO: Insert in Solvers as a classmethod.
+    if not issubclass(integrator, Integrator):
+        raise TypeError(f"The integrator {integrator} must be a subclass"
+                        " of `qutip.solver.Integrator`")
+    if not isinstance(keys, list):
+        keys = [keys]
+    for key in keys:
+        integrator_set[key] = integrator
+    if integrator.used_options and options_class:
+        for opt in integrator.used_options:
+            options_class.extra_options.add(opt)
```

Comment:
It's all very implicit that these end up being provided in `options.ode`. It would be good if that was clearer. Perhaps passing just the SolverOdeOptions to the integrator would make this better? Then at least it would be clear that the `options` received are the same class that is passed when registering the options.

Registering all the ODE options on the same `SolverOdeOptions` class also creates a namespace clash. E.g. All the zvode integrators register their options on `SolverOdeOptions` for each type of solver, but if, say, a particular integrator does not support mcsolve, then having only one `SolverOdeOptions` would mean that the option was available for mcsolve even though the integrator is not available on mcsolve.

Should we perhaps have one `SolverOdeOptions` per solver set? Doesn't necessarily have to happen in this PR.

Reply:
Changed every `SolverOptions` to `SolverOdeOptions`.

Yes, the options handling is not great and one per Solver would be better. But this can wait for now, as long as we can access the desired option.

Reply:
Sounds good for now.